e2store/src/era2.rs

```diff
@@ -0,0 +1,313 @@
+    e2store_stream: E2StoreStream,
+    storage_entries_left_to_decode: u16,
+}
+
+impl Era2 {
+    pub fn open(era2_path: &PathBuf) -> anyhow::Result<Self> {
+        let mut e2store_stream = E2StoreStream::open(era2_path)?;
+
+        let version = VersionEntry::try_from(&e2store_stream.next_entry()?)?;
+        let header = HeaderEntry::try_from(&e2store_stream.next_entry()?)?;
+
+        Ok(Self {
+            version,
+            header,
+            e2store_stream,
+            storage_entries_left_to_decode: 0,
+        })
+    }
+
+    pub fn create(era2_path: PathBuf, header: Header) -> anyhow::Result<Self> {
+        if era2_path.is_file() {
+            panic!(
+                "era2_path is not a directory, it is a file: {:?}",
+                era2_path
+            );
+        }
+        let era2_path = era2_path.join(format!(
+            "mainnet-{:010}-{}.era2",
+            header.number,
+            hex::encode(&header.state_root.as_slice()[..4])
+        ));
+        if era2_path.exists() {
+            panic!("era2 file already exists: {:?}", era2_path);
+        }
+        let mut e2store_stream = E2StoreStream::create(&era2_path)?;
+
+        let version = VersionEntry::default();
+        e2store_stream.append_entry(&version.clone().into())?;
+
+        let header = HeaderEntry { header };
+        e2store_stream.append_entry(&header.clone().try_into()?)?;
+
+        Ok(Self {
+            version,
+            header,
+            e2store_stream,
+            storage_entries_left_to_decode: 0,
+        })
+    }
+
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> anyhow::Result<Era2StreamEntry> {
+        if self.storage_entries_left_to_decode > 0 {
+            self.storage_entries_left_to_decode -= 1;
+            return Ok(Era2StreamEntry::Storage(StorageEntry::try_from(
+                &self.e2store_stream.next_entry()?,
+            )?));
+        }
+        let account = AccountEntry::try_from(&self.e2store_stream.next_entry()?)?;
+        self.storage_entries_left_to_decode = account.storage_count;
+        Ok(Era2StreamEntry::Account(account))
+    }
+
+    pub fn append_entry(&mut self, entry: &Era2StreamEntry) -> anyhow::Result<usize> {
+        let size = match entry {
+            Era2StreamEntry::Account(account) => {
+                if self.storage_entries_left_to_decode != 0 {
+                    panic!("Invalid append entry state: expected a storage entry, got an account entry. Still have {} storage entries left to append", self.storage_entries_left_to_decode);
+                }
+
+                self.storage_entries_left_to_decode = account.storage_count;
+                let entry: Entry = account.clone().try_into()?;
+                self.e2store_stream.append_entry(&entry)?;
+                entry.value.len()
+            }
+            Era2StreamEntry::Storage(storage) => {
+                if self.storage_entries_left_to_decode == 0 {
```

Comment:
nit: Should we do something like this:

```rust
match self.storage_entries_left_to_decode {
    0 => bail!(...),
    1 => ensure!(
        storage.storage.len() <= 10_000_000,
        "Storage entry can't have more than 10 million items",
    ),
    _ => ensure!(
        storage.storage.len() == 10_000_000,
        "Only last storage entry can have less than 10 million items",
    ),
}
```

Reply:
I forgot to add these checks thank you for reminding me! I really appreciate it